File: dat_learning_system-be/LMS.Backend/Data/Entities/Vocabulary.cs
namespace LMS.Backend.Data;

public class Vocabulary
{
    public Guid Id { get; set; }
    public string Word { get; set; } = string.Empty;
    public string Reading { get; set; } = string.Empty;
    public string Meaning { get; set; } = string.Empty;
    public string PartOfSpeech { get; set; } = string.Empty; // e.g., Noun, Verb
    public string JLPTLevel { get; set; } = "N5";
    public string? Explanation { get; set; }
    
    // One-to-Many Relationship
    public List<VocabularyExample> Examples { get; set; } = new();
}

public class VocabularyExample
{
    public Guid Id { get; set; }
    public string Japanese { get; set; } = string.Empty;
    public string English { get; set; } = string.Empty;
    
    public Guid VocabularyId { get; set; }
}

File: dat_learning_system-fe/src/types_interfaces/vocabulary.ts
import type { JLPTLevel } from "./kanji";

export type PartOfSpeech = "Noun" | "Verb" | "Adjective" | "Adverb" | "Particle" | "Expression" | "Other";

export interface VocabularyExample {
  id?: string;
  japanese: string; // Matches C# property name
  english: string;  // Matches C# property name
}

export interface Vocabulary {
  id: string;
  word: string;
  reading: string;
  meaning: string;
  partOfSpeech: PartOfSpeech;
  jlptLevel: JLPTLevel;
  explanation?: string;
  examples: VocabularyExample[];
}

// Used for the Upsert (Create/Update) logic
export interface UpsertVocabRequest extends Partial<Omit<Vocabulary, 'id'>> {
  id?: string;
  word: string;
  reading: string;
  meaning: string;
  partOfSpeech: PartOfSpeech;
  jlptLevel: JLPTLevel;
  examples: VocabularyExample[];
}